lib/utils/aws/s3.go

```diff
@@ -62,3 +67,57 @@ func ConvertS3Error(err error, args ...interface{}) error {
 	}
 	return err
 }
+
+// s3V2FileWriter can be used to upload data to s3 when via io.WriteCloser interface.
+type s3V2FileWriter struct {
+	// uploadFinisherErrChan is used to wait for completed upload as well as
+	// sending error message.
+	uploadFinisherErrChan <-chan error
+	pipeWriter            *io.PipeWriter
+}
+
+func NewS3V2FileWriter(ctx context.Context, s3Client *s3v2.Client, bucket, key string, uploaderOptions []func(*managerV2.Uploader), putObjectInputOptions ...func(*s3v2.PutObjectInput)) (*s3V2FileWriter, error) {
+	uploader := managerV2.NewUploader(s3Client, uploaderOptions...)
+	pr, pw := io.Pipe()
+
+	uploadParams := &s3v2.PutObjectInput{
+		Bucket: awsV2.String(bucket),
+		Key:    awsV2.String(key),
+		Body:   pr,
+	}
+
+	for _, f := range putObjectInputOptions {
+		f(uploadParams)
+	}
+	uploadFinisherErrChan := make(chan error)
+	go func() {
+		defer close(uploadFinisherErrChan)
+		_, err := uploader.Upload(ctx, uploadParams)
+		if err != nil {
+			pw.CloseWithError(err)
+		}
+		uploadFinisherErrChan <- trace.Wrap(err)
+	}()
+
+	return &s3V2FileWriter{
+		uploadFinisherErrChan: uploadFinisherErrChan,
+		pipeWriter:            pw,
+	}, nil
+}
+
+func (s *s3V2FileWriter) Write(in []byte) (int, error) {
+	bytesWritten, writeError := s.pipeWriter.Write(in)
+	if writeError != nil {
+		s.pipeWriter.CloseWithError(writeError)
+		return 0, writeError
```

Comment:
```suggestion
		return bytesWritten, writeError
```

It's generally possible to have a partial write, so this is a more conformant implementation of io.Writer.